render/gles2/renderer.c

```diff
@@ -331,6 +331,74 @@ static bool gles2_read_pixels(struct wlr_renderer *wlr_renderer,
 	return glGetError() == GL_NO_ERROR;
 }
 
+static bool gles2_blit_dmabuf(struct wlr_renderer *wlr_renderer,
+		struct wlr_dmabuf_attributes *dst_attr,
+		struct wlr_dmabuf_attributes *src_attr) {
+	assert(!wlr_renderer->rendering);
+
+	if (!gles2_procs.glEGLImageTargetRenderbufferStorageOES) {
+		return false;
+	}
+
+	bool r = false;
+	struct wlr_texture *src_tex =
+		wlr_texture_from_dmabuf(wlr_renderer, src_attr);
+
+	struct wlr_egl *egl = wlr_gles2_renderer_get_egl(wlr_renderer);
+	if (!wlr_egl_make_current(egl, EGL_NO_SURFACE, NULL)) {
+		return false;
+	}
+
+	// This is to take into account y-inversion on both buffers rather than
+	// just the source buffer.
+	bool src_inverted_y =
+		!!(src_attr->flags & WLR_DMABUF_ATTRIBUTES_FLAGS_Y_INVERT);
+	bool dst_inverted_y =
+		!!(dst_attr->flags & WLR_DMABUF_ATTRIBUTES_FLAGS_Y_INVERT);
+	struct wlr_gles2_texture *gles2_src_tex = gles2_get_texture(src_tex);
+	gles2_src_tex->inverted_y = src_inverted_y ^ dst_inverted_y;
+
+	GLuint rbo = 0;
+	glGenRenderbuffers(1, &rbo);
+	glBindRenderbuffer(GL_RENDERBUFFER, rbo);
+
+	// TODO: The imported buffer should be checked with
+	// eglQueryDmaBufModifiersEXT to see if it may be modified.
+	EGLImageKHR image = wlr_egl_create_image_from_dmabuf(egl, dst_attr);
+	gles2_procs.glEGLImageTargetRenderbufferStorageOES(GL_RENDERBUFFER, image);
+	glBindRenderbuffer(GL_RENDERBUFFER, 0);
+
+	GLuint fbo = 0;
+	glGenFramebuffers(1, &fbo);
+	glBindFramebuffer(GL_FRAMEBUFFER, fbo);
+	glFramebufferRenderbuffer(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0,
+			GL_RENDERBUFFER, rbo);
+
+	GLenum status = glCheckFramebufferStatus(GL_FRAMEBUFFER);
+	if (status != GL_FRAMEBUFFER_COMPLETE) {
+		goto out;
+	}
+
+	// TODO: use ANGLE_framebuffer_blit if available
+	float mat[9];
+	wlr_matrix_projection(mat, 1, 1, WL_OUTPUT_TRANSFORM_NORMAL);
+
+	wlr_renderer_begin(wlr_renderer, dst_attr->width, dst_attr->height);
+	wlr_renderer_clear(wlr_renderer, (float[]){ 0.0, 0.0, 0.0, 0.0 });
+	wlr_render_texture_with_matrix(wlr_renderer, src_tex, mat, 1.0f);
+	wlr_renderer_end(wlr_renderer);
+
+	r = true;
+out:
+	glBindFramebuffer(GL_FRAMEBUFFER, 0);
+	glDeleteFramebuffers(1, &fbo);
+	glDeleteRenderbuffers(1, &rbo);
+	wlr_egl_destroy_image(egl, image);
+	wlr_texture_destroy(src_tex);
+	wlr_egl_unset_current(egl);
```

Comment:
As discussed on IRC, this causes issues because this function is called in wlr_output's precommit event handler. Unsetting the current EGL context makes `wlr_egl_swap_buffers` fail.

A workaround would be to leave the EGL context behind here (by _not_ calling `wlr_egl_unset_current`). A proper solution would involve restoring the previous context or ref-counting the number of times `wlr_egl_make_current` is called (this will probably mess up the current EGL surface).

Reply:
In order to restore the context, I would have to save it before calling `blit_dmabuf()` and then restore it after. Is there an established pattern for that?

Reply:
Maybe something like this?
```
diff --git a/render/gles2/renderer.c b/render/gles2/renderer.c
index 8c5651db..a3ac3ec0 100644
--- a/render/gles2/renderer.c
+++ b/render/gles2/renderer.c
@@ -340,6 +340,9 @@ static bool gles2_blit_dmabuf(struct wlr_renderer *wlr_renderer,
                return false;
        }

+       EGLContext old_ctx = eglGetCurrentContext();
+       EGLSurface old_surf = eglGetCurrentSurface(EGL_DRAW);
+
        bool r = false;
        struct wlr_texture *src_tex =
                wlr_texture_from_dmabuf(wlr_renderer, src_attr);
@@ -395,7 +398,7 @@ out:
        glDeleteRenderbuffers(1, &rbo);
        wlr_egl_destroy_image(egl, image);
        wlr_texture_destroy(src_tex);
-       wlr_egl_unset_current(egl);
+       wlr_egl_make_current(egl, old_ctx, old_surf);
        return r;
 }
```

Reply:
>Is there an established pattern for that?

No, because we overwrite the previous context everywhere right now.

>Maybe something like this?

Something like that should work, but maybe also extract the `EGL_READ` surface separately to avoid surprises. And also use `eglGetCurrentDisplay` in case we're juggling between multiple displays (like we do in the DRM backend).

Reply:
Maybe this should be a pair of functions called `wlr_egl_save_current(struct wlr_egl_context*)` and `wlr_egl_restore_current(struct wlr_egl_context*)`?

Reply:
Push and pop?

Reply:
push/pop would be fancy, but I believe the call stack will suffice.

Reply:
I have now added the utility functions as discussed on IRC and applied them to `blit_dmabuf()`.